### PR TITLE
Aa dc sort list

### LINKED ISSUE #12: 12. As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon.
Possible item states are as follows:

* Need to buy soon (fewer than 7 days)
* Need to buy kind of soon (between 7 & 30 days)
* Need to buy not soon (more than 30 days)
* Inactive (when there’s only 1 purchase in the database or the purchase is really out of date [the time that has elapsed since the last purchase is 2x what was estimated])

AC:

* Items in the list are shown as visually distinct (e.g., with a different background color on the list item) according to how soon the item is expected to be bought again: Soon, Kind of soon, Not soon, Inactive
* Items should be sorted by the estimated number of days until the next purchase
* Items with the same number of estimated days until next purchase should be sorted alphabetically
* Items in the different states should be described distinctly when read by a screen reader

Notes:

* The criteria for an item being inactive has tripped up teams in the past. It's been re-worded to be more clear, but the gist is that an item is inactive if the time that has elapsed since its last purchase is twice (or more) what was estimated the last time it was purchased.
* E.g., if the last estimate was 14 days and 28 days have elapsed, the item is considered inactive.
* Teams in the past have also overengineered the solution here by breaking the list out into separate groups, but it should be achievable in a single list where the visual treatment is different based on what category the item falls into.
* Be sure to not only use color to denote how soon the user will need to buy the item, as noted in the wireframe. https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/list-populated.png
* Good luck, you got this!

### PR DESCRIPTION
**Pull request by adiati98**
## Description

This PR adds several functionalities:
- Sorting items by the estimated number of days until the next purchase.
- Sorting items with the same number of estimated days until next purchase alphabetically.
- Add background color to each item based on their `nextPurchase` days for visual distinction.
- Add `aria-label` to each item based on their `nextPurchase` days for accessibility purpose. 

## Related Issue

Closes #12 

## Acceptance Criteria

- Items in the list are shown as visually distinct (e.g., with a different background color on the list item) according to how soon the item is expected to be bought again: Soon, Kind of soon, Not soon, Inactive
- Items should be sorted by the estimated number of days until the next purchase
- Items with the same number of estimated days until the next purchase should be sorted alphabetically
- Items in the different states should be described distinctly when read by a screen reader

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

lis-item-before

### After

list-item

## Testing Steps / QA Criteria

- Pull the branch.
- Run npm start
- Launch the browser to preview at http://localhost:3000/
- Enter a token from Firebase to join the existing list or create a new list and add several items.
  By now we should already see the background color for each item based on their `nextPurchase`.
   - Green -> 7 days
   - Yellow -> 14 days
   - Purple -> 30 days
- Open browser's devtools.
- Inspect the `<span>` inside the `<label>` of `input checkbox` to check the `aria-label`.
  Here we will see something like  `<span class="hide-span" aria-label="Buy Chocolate every 14 days">Chocolate</span>`.
- Go to Firebase.
- Rollback the date of `lastPurchase` on an item.
- Go to our application and check that item again.
  The expected behavior is that the list is now sorted based on the `estimatedPurchaseInterval`. And items with the same `estimatedPurchaseInterval` are sorted alphabetically.

### CODE REVIEW

**Comment by edemaukabi** [comment]
This looks good to me

File: src/components/ListItem.js [hunk 1]
```diff
@@ -6,26 +6,26 @@ import { collection, doc, onSnapshot, updateDoc } from 'firebase/firestore';
...
   const [error, setError] = useState(false);
   const [emptyList, setEmptyList] = useState(false);
-  const [renderInput, setRenderInput] = useState(false);
   const [searchlist, setSearchList] = useState('');
 
   const currToken = localStorage.getItem('currToken');
   let currentCollectionRef;
   const itemsCollectionRef = collection(db, 'shopping-list');
   const history = useHistory();
+  const DayInMilliSeconds = 60 * 60 * 24 * 1000;
```
Comment (faysvas) [inline src/components/ListItem.js:22]: Great that you don't have a magic number any more!

File: src/components/ListItem.js [hunk 2]
```diff
@@ -76,20 +75,87 @@ const ListItem = () => {
...
   };
 
+  const addBgColor = (nextPurchase) => {
+    if (nextPurchase === 7) {
+      return 'green-bg';
+    } else if (nextPurchase === 14) {
+      return 'yellow-bg';
+    } else {
+      return 'purple-bg';
+    }
```
Comment (faysvas) [inline src/components/ListItem.js:85]: Here an color for the inactive state should be added too

File: src/components/ListItem.js [hunk 3]
```diff
@@ -76,20 +75,87 @@ const ListItem = () => {
...
+  };
+
+  const addAriaLabel = (nextPurchase, itemName) => {
+    if (itemName && nextPurchase === 7) {
+      return `Buy ${itemName} every ${nextPurchase}} days`;
```
Comment (faysvas) [inline src/components/ListItem.js:90]: There's a typo here, there's an extra "}" at the end of `${nextPurchase}}`

File: src/components/ListItem.js [hunk 4]
```diff
@@ -76,20 +75,87 @@ const ListItem = () => {
...
+    } else if (itemName && nextPurchase === 14) {
+      return `Buy ${itemName} every ${nextPurchase}} days`;
```
Comment (faysvas) [inline src/components/ListItem.js:92]: Same here

File: src/components/ListItem.js [hunk 5]
```diff
@@ -76,20 +75,87 @@ const ListItem = () => {
...
+    } else if (itemName && nextPurchase === 30) {
+      return `Buy ${itemName} every ${nextPurchase}} days`;
```
Comment (faysvas) [inline src/components/ListItem.js:94]: Same here

File: src/components/ListItem.js [hunk 6]
```diff
@@ -76,20 +75,87 @@ const ListItem = () => {
...
+      if (a.estimatedPurchaseInterval < b.estimatedPurchaseInterval) {
+        return -1;
+      } else if (a.estimatedPurchaseInterval > b.estimatedPurchaseInterval) {
+        return 1;
+      } else {
+        if (a.itemName < b.itemName) {
+          return -1;
+        } else {
+          return 1;
+        }
```
Comment (faysvas) [inline src/components/ListItem.js:135]: Since this code is repeated twice, this code could be in a function to avoid code duplication

**Review by Amaka202 — COMMENTED** [review]
Hi team! well done for working so hard on this issue!💪

I checked out the updated version. Please check below.

The list is indeed sorted according to when I will need them again but only when I have bought the items more than once. But items I am buying for the first time kinda spoils everything. So the user story says.

```
As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon
```

Screenshot 2021-11-19 at 11 53 23 AM

The screenshot above fulfils that. But when I add items for the first time, I am no longer able to know which items are due next. see screenshot below.

Screenshot 2021-11-19 at 12 03 38 PM

As you can see above. I have just added new items to my list so my app can help me remember when to buy them again (remember I was asked when I am likely to buy again so the app can remind me). But I can't say from the list view which items are due first, do I follow the colours or do I follow the order in which they appear. Remember I didn't build the app, I don't know what’s happening under the hood. I am just a user who found this app to help me keep track of my shopping list and show me when each item is due. This is literally the essence of the entire project - to tell/show a user which items are due first and when.

See the user story once more.

```
As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon
```

see attached wireframe below for reference. in the wireframe,  the items belonging to the same category have the same colour, whether checked or not.

Screenshot 2021-11-19 at 12 35 30 PM

I would suggest the approach I suggested already on slack, let `estimatedPurchaseInterval` and `nextPurchase` be merged into one value. That way you are sorting against one value. But sorting with both could also work because the two values aren't supposed to be significantly different. Whichever way you decide to go is fine👍👍

Also, please @edemaukabi  and @fatima-ola  could you guys join hands with @danachen and @adiati98 to get the issue working. Remember it's a team project and it will be faster if everyone joins ideas together to get it done😎😎

You can also chat us on slack should you need any help.❤️

**Review by Amaka202 — APPROVED** [review]

**Pull request merged**